upstairs/src/test.rs

```diff
@@ -5105,4 +5245,970 @@ mod up_test {
+        assert_eq!(jobs[1].work.deps(), &[jobs[0].ds_id]);
+        assert_eq!(jobs[2].work.deps(), &[jobs[1].ds_id]);
+    }
+
+    #[tokio::test]
+    async fn test_deps_flushes_depend_on_flushes_and_all_writes() {
+        // Test that the following job dependency graph is made:
+        //
+        // flush -> write -> flush -> write -> flush
+        //       -> write ->       -> write ->
+        //       ---------->       -> write ->
+        //                         ---------->
+        //       block
+        // op# | 0 1 2 | deps
+        // ----|-------|-----
+        //   0 | FFFFF |
+        //   1 | W     | 0
+        //   2 |   W   | 0
+        //   3 | FFFFF | 0,1,2
+        //   4 | W     | 3
+        //   5 |   W   | 3
+        //   6 |     W | 3
+        //   7 | FFFFF | 3,4,5,6
+
+        let upstairs = make_upstairs();
+        upstairs.set_active().await.unwrap();
+
+        // op 0
+        upstairs.submit_flush(None, None).await.unwrap();
+
+        // ops 1 to 2
+        for i in 0..2 {
+            upstairs
+                .submit_write(
+                    Block::new_512(i),
+                    Bytes::from(vec![0xff; 512]),
+                    None,
+                    false,
+                )
+                .await
+                .unwrap();
+        }
+
+        // op 3
+        upstairs.submit_flush(None, None).await.unwrap();
+
+        // ops 4 to 6
+        for i in 0..3 {
+            upstairs
+                .submit_write(
+                    Block::new_512(i),
+                    Bytes::from(vec![0xff; 512]),
+                    None,
+                    false,
+                )
+                .await
+                .unwrap();
+        }
+
+        // op 7
+        upstairs.submit_flush(None, None).await.unwrap();
+
+        let ds = upstairs.downstairs.lock().await;
+        let keys: Vec<&u64> = ds.active.keys().sorted().collect();
+        let jobs: Vec<&DownstairsIO> =
+            keys.iter().map(|k| ds.active.get(k).unwrap()).collect();
+        assert_eq!(jobs.len(), 8);
+
+        assert!(jobs[0].work.deps().is_empty()); // flush (op 0)
+
+        assert_eq!(jobs[1].work.deps(), &vec![jobs[0].ds_id]); // write (op 1)
+        assert_eq!(jobs[2].work.deps(), &vec![jobs[0].ds_id]); // write (op 2)
+
+        assert_eq!(
+            hashset(jobs[3].work.deps()), // flush (op 3)
+            hashset(&[jobs[0].ds_id, jobs[1].ds_id, jobs[2].ds_id]),
+        );
+
+        assert_eq!(jobs[4].work.deps(), &[jobs[3].ds_id]); // write (op 4)
+        assert_eq!(jobs[5].work.deps(), &[jobs[3].ds_id]); // write (op 5)
+        assert_eq!(jobs[6].work.deps(), &[jobs[3].ds_id]); // write (op 6)
+
+        assert_eq!(
+            hashset(jobs[7].work.deps()), // flush (op 7)
+            hashset(&[
+                jobs[3].ds_id,
+                jobs[4].ds_id,
+                jobs[5].ds_id,
+                jobs[6].ds_id
+            ]),
+        );
+    }
+
```

Comment:
Some more tests spanning extents with operations depending on things
from the previous extent.
```
//       block   block
// op# | 0 1 2 | 0 1 2 | deps
// ----|-------|-------|-----
//   0 | W W   |       |
//   1 |   W W | W W W | 0
//   2 |       |   W   | 1? or 0,1?
```

Or, there and back:
```
//       block   block
// op# | 0 1 2 | 0 1 2 | deps
// ----|-------|-------|-----
//   0 | W W   |       |
//   1 |   W W | W W W | 0
//   2 |       |   W   |
//   3 |     W | W W   |
//   4 |     R |       |
```

IOs from another mother:
```
//       block   block
// op# | 0 1 2 | 0 1 2 | deps
// ----|-------|-------|-----
//   0 | W W   |       |
//   1 |       |   W W | 0
//   2 |   W W | W W   | 0, 1
```

Reply:
done in c9faeeb